# FIFA World Cup Tournament

This project focuses on analyzing the FIFA World Cup Tournament. The dataset used in this project is from Kaggle. The dataset contains information about the FIFA World Cup matches.

Our goal is to analyze the dataset and answer the following questions:
1. Do teams perform better when playing as the host team in the men's FIFA World Cup?
2. Who were the top goal scorers, individually and by team?
3. Are the teams who perform better in the first (1st) or second (2nd) half of the match more likely to win the match?

## Importing Libraries + Cleaning Data

First, we will import the necessary libraries and clean the data.

### Import Libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np

In [2]:
# Datasets URL
results_url = 'https://raw.githubusercontent.com/cogxen/datasets/main/fifa-world-cup-tournament/results.csv'
goalscorers_url = 'https://raw.githubusercontent.com/cogxen/datasets/main/fifa-world-cup-tournament/goalscorers.csv'
shootouts_url = 'https://raw.githubusercontent.com/cogxen/datasets/main/fifa-world-cup-tournament/shootouts.csv'

### Load and Inspect the Datasets

This FIFA World Cup matches data is stored in three (3) different CSV files:
- [results.csv](https://github.com/cogxen/datasets/blob/main/for-laboratory/projects/results.csv)
- [goalscorers.csv](https://github.com/cogxen/datasets/blob/main/for-laboratory/projects/goalscorers.csv)
- [shootouts.csv](https://github.com/cogxen/datasets/blob/main/for-laboratory/projects/shootouts.csv)

We need to load these datasets and merge them accordingly. 

First, let's load the `results.csv` dataset.

In [3]:
# Load the results.csv dataset
df1 = pd.read_csv(results_url)
df1.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  Scotland   England           0           0   Friendly  Glasgow   
1  1873-03-08   England  Scotland           4           2   Friendly   London   
2  1874-03-07  Scotland   England           2           1   Friendly  Glasgow   
3  1875-03-06   England  Scotland           2           2   Friendly   London   
4  1876-03-04  Scotland   England           3           0   Friendly  Glasgow   

    country  neutral  
0  Scotland    False  
1   England    False  
2  Scotland    False  
3   England    False  
4  Scotland    False

In [4]:
# Show results.csv dataset information
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44557 entries, 0 to 44556
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        44557 non-null  object
 1   home_team   44557 non-null  object
 2   away_team   44557 non-null  object
 3   home_score  44557 non-null  int64 
 4   away_score  44557 non-null  int64 
 5   tournament  44557 non-null  object
 6   city        44557 non-null  object
 7   country     44557 non-null  object
 8   neutral     44557 non-null  bool  
dtypes: bool(1), int64(2), object(6)
memory usage: 2.8+ MB


The `results.csv` dataset contains the following columns:

- `date`: The date the match was played
- `home_team`: The home team
- `away_team`: The away team
- `home_score`: The home team's score
- `away_score`: The away team's score
- `tournament`: The tournament the match was played in
- `city`: The city the match was played in
- `country`: The country the match was played in
- `neutral`: Whether the match was played in a neutral location

Next, let's load the `goalscorers.csv` dataset.

In [5]:
# Load the goalscorers.csv dataset
df2 = pd.read_csv(goalscorers_url)
df2.head()

date  home_team away_team       team            scorer  minute  \
0  1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    44.0   
1  1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    55.0   
2  1916-07-02      Chile   Uruguay    Uruguay  Isabelino Gradín    70.0   
3  1916-07-02      Chile   Uruguay    Uruguay   José Piendibene    75.0   
4  1916-07-06  Argentina     Chile  Argentina     Alberto Ohaco     2.0   

   own_goal  penalty  
0     False    False  
1     False    False  
2     False    False  
3     False    False  
4     False    False

In [6]:
# Show goalscorers.csv dataset information
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41008 entries, 0 to 41007
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date       41008 non-null  object 
 1   home_team  41008 non-null  object 
 2   away_team  41008 non-null  object 
 3   team       41008 non-null  object 
 4   scorer     40959 non-null  object 
 5   minute     40750 non-null  float64
 6   own_goal   41008 non-null  bool   
 7   penalty    41008 non-null  bool   
dtypes: bool(2), float64(1), object(5)
memory usage: 2.0+ MB


The `goalscorers.csv` dataset contains the following columns:

- `date`: The date the match was played
- `home_team`: The home team
- `away_team`: The away team
- `team`: The team the goal scorer played for
- `scorer`: The goal scorer
- `minute`: The minute the goal was scored
- `own_goal`: Whether the goal was an own goal
- `penalty`: Whether the goal was a penalty

Finally, let's load the `shootouts.csv` dataset.

In [7]:
# Load the shootouts.csv dataset
df3 = pd.read_csv(shootouts_url)
df3.head()

date    home_team         away_team       winner
0  1967-08-22        India            Taiwan       Taiwan
1  1971-11-14  South Korea  Vietnam Republic  South Korea
2  1972-05-07  South Korea              Iraq         Iraq
3  1972-05-17     Thailand       South Korea  South Korea
4  1972-05-19     Thailand          Cambodia     Thailand

In [8]:
# Show shootouts.csv dataset information
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 547 entries, 0 to 546
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   date       547 non-null    object
 1   home_team  547 non-null    object
 2   away_team  547 non-null    object
 3   winner     547 non-null    object
dtypes: object(4)
memory usage: 17.2+ KB


The `shootouts.csv` dataset contains the following columns:

- `date`: The date the match was played
- `home_team`: The home team
- `away_team`: The away team
- `winner`: The winner of the shootout

Now that we have loaded the datasets, we can proceed to clean the data.

### Data Cleaning and Preparation

Before we can merge `df1` (results.csv) and `df3` (shootouts.csv) datasets into a single dataset, we need to clean first the column names and row values in both datasets. We want to make sure that all column names and in lower case and have no whitespaces also in any of the text data sice we will be merging on those columns.

Alright, let's create a function that can be applied to both datasets to clean the column names and row values.

In [9]:
# Function to clean column names and row values
def clean_data(df):
    df.columns = df.columns.str.lower()
    df.columns = df.columns.str.strip()
    
    for column in df.columns:
        if df[column].dtype == 'object':
            df[column] = df[column].str.lower()
            df[column] = df[column].str.strip()
            
    return df

Now, let's apply the `clean_data` function to both `df1` and `df3` datasets.

In [10]:
# Clean the results.csv dataset
results = clean_data(df1)

# Show cleaned results.csv dataset
results.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  scotland   england           0           0   friendly  glasgow   
1  1873-03-08   england  scotland           4           2   friendly   london   
2  1874-03-07  scotland   england           2           1   friendly  glasgow   
3  1875-03-06   england  scotland           2           2   friendly   london   
4  1876-03-04  scotland   england           3           0   friendly  glasgow   

    country  neutral  
0  scotland    False  
1   england    False  
2  scotland    False  
3   england    False  
4  scotland    False

In [11]:
# Clean the shootouts.csv dataset
shootouts = clean_data(df3)

# Show cleaned shootouts.csv dataset
shootouts.head()

date    home_team         away_team       winner
0  1967-08-22        india            taiwan       taiwan
1  1971-11-14  south korea  vietnam republic  south korea
2  1972-05-07  south korea              iraq         iraq
3  1972-05-17     thailand       south korea  south korea
4  1972-05-19     thailand          cambodia     thailand

Now that we have cleaned the `results.csv` and `shootouts.csv` datasets, we can proceed to merge them into a single dataset. We will do this by matching the `date`, `home_team`, and `away_team` columns in both datasets. And lastly, let's specify a **left** join to return all the matches in `results`, whether they ended in a shootout or not.

Don't forget to show the first five (5) rows of the merged dataset and its information.

In [12]:
# Merge the results and shootouts datasets
results_shootouts = pd.merge(results, shootouts, on=['date', 'home_team', 'away_team'], how='left')

# Show the merged dataset
results_shootouts.head()

date home_team away_team  home_score  away_score tournament     city  \
0  1872-11-30  scotland   england           0           0   friendly  glasgow   
1  1873-03-08   england  scotland           4           2   friendly   london   
2  1874-03-07  scotland   england           2           1   friendly  glasgow   
3  1875-03-06   england  scotland           2           2   friendly   london   
4  1876-03-04  scotland   england           3           0   friendly  glasgow   

    country  neutral winner  
0  scotland    False    NaN  
1   england    False    NaN  
2  scotland    False    NaN  
3   england    False    NaN  
4  scotland    False    NaN

In [13]:
# Show the merged dataset information
results_shootouts.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 44557 entries, 0 to 44556
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        44557 non-null  object
 1   home_team   44557 non-null  object
 2   away_team   44557 non-null  object
 3   home_score  44557 non-null  int64 
 4   away_score  44557 non-null  int64 
 5   tournament  44557 non-null  object
 6   city        44557 non-null  object
 7   country     44557 non-null  object
 8   neutral     44557 non-null  bool  
 9   winner      546 non-null    object
dtypes: bool(1), int64(2), object(7)
memory usage: 3.4+ MB


As you can see, the `winner` column only contains data on matches that ended in a shootout. The **left** join inserts `NaN` values for all matches/games that ended without any penalty shootout. Later in the data cleaning and preparation process, we will address this issue.

Now, let's analyze the result of the FIFA World Cup tournament. Let's display the top 10 most frequent tournaments in the dataset.

In [14]:
# Display the top 10 most frequent tournaments
results_shootouts['tournament'].value_counts().head(10)

friendly                                17593
fifa world cup qualification             7878
uefa euro qualification                  2631
african cup of nations qualification     1976
fifa world cup                            964
copa américa                              841
afc asian cup qualification               764
african cup of nations                    741
cecafa cup                                620
cfu caribbean cup qualification           606
Name: tournament, dtype: int64

As you can see, there are **964** matches in the dataset that are part of the FIFA World Cup tournament.

Now, let's move on to the analysis of the FIFA World Cup tournament. Filter the dataset to only include matches that are part of the FIFA World Cup tournament.

In [15]:
# Filter the dataset to include only FIFA World Cup matches
fifa_wc_mask = results_shootouts['tournament'] == 'fifa world cup'
fifa_wc = results_shootouts[fifa_wc_mask].copy()

# Show the first five (5) rows of the FIFA World Cup dataset
fifa_wc.head()

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    belgium  united states           0           3   
1312  1930-07-13     france         mexico           4           1   
1313  1930-07-14     brazil     yugoslavia           1           2   
1314  1930-07-14       peru        romania           1           3   
1315  1930-07-15  argentina         france           1           0   

          tournament        city  country  neutral winner  
1311  fifa world cup  montevideo  uruguay     True    NaN  
1312  fifa world cup  montevideo  uruguay     True    NaN  
1313  fifa world cup  montevideo  uruguay     True    NaN  
1314  fifa world cup  montevideo  uruguay     True    NaN  
1315  fifa world cup  montevideo  uruguay     True    NaN

In [16]:
# Show the FIFA World Cup dataset information
fifa_wc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 964 entries, 1311 to 44358
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        964 non-null    object
 1   home_team   964 non-null    object
 2   away_team   964 non-null    object
 3   home_score  964 non-null    int64 
 4   away_score  964 non-null    int64 
 5   tournament  964 non-null    object
 6   city        964 non-null    object
 7   country     964 non-null    object
 8   neutral     964 non-null    bool  
 9   winner      35 non-null     object
dtypes: bool(1), int64(2), object(7)
memory usage: 76.3+ KB


As you can see, **35** out of **964** matches have ended in penalty shootouts. One reason that this numer is relatively low is because that's how the FIFA World Cup tournament is structured. The tournament has a knockout stage where matches must have a winner, which is why penalty shootouts are used to determine the winner if the match ends in a draw. 

World Cup tournaments start with **group play**, where teams play in small round-robin setup/tournaments that allows ties. The knockout stage is where penalty shootouts are used to determine the winner if the match ends in a draw.

Now, we need to know the number of winners that aren't result of penalty shootouts. Let's create some columns that will help us to fill in those values and it will be useful in other aspects of our analysis later on. 

Let's create a new columns:
- `total_goals`: The total number of goals scored in the match by both teams
- `win_margin`: This where will be stored the margin of victory by `home_score` - `away_score`
    - positive value means the home team won
    - negative value means the away team won
    - if `win_margin` is `0`, the match ended in a draw or a match that ended in a penalty shootout 

In [17]:
# Create a new column 'total_goals' that contains the total number of goals scored in the match
fifa_wc['total_goals'] = fifa_wc['home_score'] + fifa_wc['away_score']

# Create a new column 'win_margin' that contains the margin of victory
fifa_wc['win_margin'] = fifa_wc['home_score'] - fifa_wc['away_score']

# Show the first five (5) rows of the FIFA World Cup dataset
fifa_wc.head()

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    belgium  united states           0           3   
1312  1930-07-13     france         mexico           4           1   
1313  1930-07-14     brazil     yugoslavia           1           2   
1314  1930-07-14       peru        romania           1           3   
1315  1930-07-15  argentina         france           1           0   

          tournament        city  country  neutral winner  total_goals  \
1311  fifa world cup  montevideo  uruguay     True    NaN            3   
1312  fifa world cup  montevideo  uruguay     True    NaN            5   
1313  fifa world cup  montevideo  uruguay     True    NaN            3   
1314  fifa world cup  montevideo  uruguay     True    NaN            4   
1315  fifa world cup  montevideo  uruguay     True    NaN            1   

      win_margin  
1311          -3  
1312           3  
1313          -1  
1314          -2  
1315           1

As we notice, the `winner` column only contains the winners of matches that resulted in a penalty shootout. We need to rename this column to `shootout_winner` .

In [18]:
# Rename the 'winner' column to 'shootout_winner'
fifa_wc.rename(columns={'winner': 'shootout_winner'}, inplace=True)

# Show the first five (5) rows of the FIFA World Cup dataset
fifa_wc.head()

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    belgium  united states           0           3   
1312  1930-07-13     france         mexico           4           1   
1313  1930-07-14     brazil     yugoslavia           1           2   
1314  1930-07-14       peru        romania           1           3   
1315  1930-07-15  argentina         france           1           0   

          tournament        city  country  neutral shootout_winner  \
1311  fifa world cup  montevideo  uruguay     True             NaN   
1312  fifa world cup  montevideo  uruguay     True             NaN   
1313  fifa world cup  montevideo  uruguay     True             NaN   
1314  fifa world cup  montevideo  uruguay     True             NaN   
1315  fifa world cup  montevideo  uruguay     True             NaN   

      total_goals  win_margin  
1311            3          -3  
1312            5           3  
1313            3          -1  
1314            4          -2  
1315            1           1

Now, let's re-create the `winner` column to determine the winners of each match that did't end in a penalty shootout and without any missing value (`NaN`) like before. 

We will use the `shootout_winner` and `win_margin` columns to determine the winner of each match. We need to handle the `NaN` values in the `shootout_winner` column by converting the column to `string` values and use the string comparison method to test if each value is `NaN`. 

Pandas has a built it method which is `pd.isna()` that can be used to test if a value is `NaN`.

In [19]:
# Determine the winner of each match
def determine_winner(row):
    if not pd.isna(row['shootout_winner']):
        winner = row['shootout_winner']
        
    elif row['win_margin'] > 0:
        winner = row['home_team']
        
    elif row['win_margin'] < 0:
        winner = row['away_team']
        
    else:
        winner = 'draw'
        
    return winner

In [20]:
# Apply the determine_winner function to the FIFA World Cup dataset
fifa_wc['winner'] = fifa_wc.apply(determine_winner, axis=1)

# Show the first five (5) rows of the FIFA World Cup dataset
fifa_wc.head()

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    belgium  united states           0           3   
1312  1930-07-13     france         mexico           4           1   
1313  1930-07-14     brazil     yugoslavia           1           2   
1314  1930-07-14       peru        romania           1           3   
1315  1930-07-15  argentina         france           1           0   

          tournament        city  country  neutral shootout_winner  \
1311  fifa world cup  montevideo  uruguay     True             NaN   
1312  fifa world cup  montevideo  uruguay     True             NaN   
1313  fifa world cup  montevideo  uruguay     True             NaN   
1314  fifa world cup  montevideo  uruguay     True             NaN   
1315  fifa world cup  montevideo  uruguay     True             NaN   

      total_goals  win_margin         winner  
1311            3          -3  united states  
1312            5           3         france  
1313            3          -1     yugoslavia  
1314            4          -2        romania  
1315            1           1      argentina

In [21]:
# Show the FIFA World Cup dataset information
fifa_wc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 964 entries, 1311 to 44358
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   date             964 non-null    object
 1   home_team        964 non-null    object
 2   away_team        964 non-null    object
 3   home_score       964 non-null    int64 
 4   away_score       964 non-null    int64 
 5   tournament       964 non-null    object
 6   city             964 non-null    object
 7   country          964 non-null    object
 8   neutral          964 non-null    bool  
 9   shootout_winner  35 non-null     object
 10  total_goals      964 non-null    int64 
 11  win_margin       964 non-null    int64 
 12  winner           964 non-null    object
dtypes: bool(1), int64(4), object(8)
memory usage: 98.8+ KB


Now let's create new two additional columns using `winner` column.

- `home_win`: A boolean column that indicates if the home team won the match, return `True` if the `winner` is the `home_team` and `False` otherwise
- `away_win`: A boolean column that indicates if the away team won the match, return `True` if the `winner` is the `away_team` and `False` otherwise

In [22]:
# Create a new column 'home_win' that indicates if the home team won the match
fifa_wc['home_win'] = fifa_wc['winner'] == fifa_wc['home_team']

# Create a new column 'away_win' that indicates if the away team won the match
fifa_wc['away_win'] = fifa_wc['winner'] == fifa_wc['away_team']

# Show the first five (5) rows of the FIFA World Cup dataset
fifa_wc.head()

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    belgium  united states           0           3   
1312  1930-07-13     france         mexico           4           1   
1313  1930-07-14     brazil     yugoslavia           1           2   
1314  1930-07-14       peru        romania           1           3   
1315  1930-07-15  argentina         france           1           0   

          tournament        city  country  neutral shootout_winner  \
1311  fifa world cup  montevideo  uruguay     True             NaN   
1312  fifa world cup  montevideo  uruguay     True             NaN   
1313  fifa world cup  montevideo  uruguay     True             NaN   
1314  fifa world cup  montevideo  uruguay     True             NaN   
1315  fifa world cup  montevideo  uruguay     True             NaN   

      total_goals  win_margin         winner  home_win  away_win  
1311            3          -3  united states     False      True  
1312            5           3         france      True     False  
1313            3          -1     yugoslavia     False      True  
1314            4          -2        romania     False      True  
1315            1           1      argentina      True     False

As you can see, our dataset is now contains a lot of new information, such as the total number of goals scored in each match, the margin of victory, and the winner of each match. Now let's clean up the dataset by dropping redundant columns.

Let's drop the following columns:

- `tournament`: We are only analyzing FIFA World Cup matches
- `city`: We are only interested in the country where the match was played
- `shootout_winner`: We have already determined the winner of each match

In [23]:
# Drop the 'tournament', 'city', and 'shootout_winner' columns
fifa_wc.drop(columns=['tournament', 'city', 'shootout_winner'], inplace=True)

# Show the first five (5) rows of the FIFA World Cup dataset
fifa_wc.head()

date  home_team      away_team  home_score  away_score  country  \
1311  1930-07-13    belgium  united states           0           3  uruguay   
1312  1930-07-13     france         mexico           4           1  uruguay   
1313  1930-07-14     brazil     yugoslavia           1           2  uruguay   
1314  1930-07-14       peru        romania           1           3  uruguay   
1315  1930-07-15  argentina         france           1           0  uruguay   

      neutral  total_goals  win_margin         winner  home_win  away_win  
1311     True            3          -3  united states     False      True  
1312     True            5           3         france      True     False  
1313     True            3          -1     yugoslavia     False      True  
1314     True            4          -2        romania     False      True  
1315     True            1           1      argentina      True     False

In [24]:
# Show the FIFA World Cup dataset information
fifa_wc.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 964 entries, 1311 to 44358
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         964 non-null    object
 1   home_team    964 non-null    object
 2   away_team    964 non-null    object
 3   home_score   964 non-null    int64 
 4   away_score   964 non-null    int64 
 5   country      964 non-null    object
 6   neutral      964 non-null    bool  
 7   total_goals  964 non-null    int64 
 8   win_margin   964 non-null    int64 
 9   winner       964 non-null    object
 10  home_win     964 non-null    bool  
 11  away_win     964 non-null    bool  
dtypes: bool(3), int64(4), object(5)
memory usage: 78.1+ KB


Now that we have cleaned the dataset, we can proceed to analyze the data and answer the questions we have set out to answer.

## Exploratory Data Analysis

### Question 1. Do teams perform better when playing as the host team in?

If you'll notice, the home team in our dataset is slightly misleading. In the World Cup, one team is always designated as the home team even if they are not playing in their home country. This is because the home team is typically listed first in the match-up. But what we really wanted to know is whether teams perform better when playing in their home country or not. 

In some ways, we can tell that the designated home team is also the host team for that match. Matches where `home_team` and `country` are the same are considered home matches and matches where `neutral` is `False` are considered.

First, let's count how many matches were played where the home team is also the host team.

In [25]:
# Count the number of matches where the home team is also the host team
fifa_wc['neutral'].value_counts()

True     843
False    121
Name: neutral, dtype: int64

As you can see, there are **843** matches where the home team is also the host team.

But again, we need to remember that the home team is not always the host team. We need to count the number of matches where the home team is playing in their home country.

Now, let's split the FIFA World Cup matches for matches played by host team.

In [26]:
# Split the FIFA World Cup matches for matches played by host team
host_team_mask = fifa_wc['home_team'] == fifa_wc['country']
host_team = fifa_wc[host_team_mask]

# Show the first five (5) rows of the host team dataset
host_team.head()

date home_team      away_team  home_score  away_score  country  \
1320  1930-07-18   uruguay           peru           1           0  uruguay   
1325  1930-07-21   uruguay        romania           4           0  uruguay   
1329  1930-07-27   uruguay     yugoslavia           6           1  uruguay   
1330  1930-07-30   uruguay      argentina           4           2  uruguay   
1694  1934-05-27     italy  united states           7           1    italy   

      neutral  total_goals  win_margin   winner  home_win  away_win  
1320    False            1           1  uruguay      True     False  
1325    False            4           4  uruguay      True     False  
1329    False            7           5  uruguay      True     False  
1330    False            6           2  uruguay      True     False  
1694    False            8           6    italy      True     False

Let's now compute the summary statistics for the matches played by the host team.

In [27]:
# Compute the summary statistics for matches played by the host team
host_team.describe()

home_score  away_score  total_goals  win_margin
count  121.000000  121.000000   121.000000  121.000000
mean     1.834711    0.991736     2.826446    0.842975
std      1.561768    1.268174     2.023519    2.000034
min      0.000000    0.000000     0.000000   -6.000000
25%      1.000000    0.000000     1.000000    0.000000
50%      2.000000    1.000000     3.000000    1.000000
75%      2.000000    1.000000     4.000000    2.000000
max      7.000000    7.000000    12.000000    6.000000

In [28]:
# Display the number of unique home teams
host_team['home_team'].nunique()

18

In [29]:
# Show the average win of the host team
host_team['home_win'].mean()

0.6528925619834711

As you can see, there are **18** unique home teams in the dataset that won about **65.3%** of their matches and with an average of **0.843** more goals that their opponents.

Now let's compare their win record as host team to their win record as a non-host team.

Let's start by calculating the percentage of the host team's wins as the host team.

In [30]:
# Calculate the percentage of the host team's wins as the host team
host_team_win_percentage = host_team.groupby('home_team')['home_win'].mean().reset_index()
host_team_win_percentage.columns = ['team', 'win_percentage']

# Show the first five (5) rows of the host team win percentage
host_team_win_percentage.head()

team  win_percentage
0  argentina        0.714286
1     brazil        0.615385
2      chile        0.666667
3    england        0.833333
4     france        0.888889

Now let's calculate the percentage of the host team's wins as the non-host team.

To make ths calculation, we need to perform a **Split-Apply-Combine** operation. We will split the dataset into two groups: matches where the home team is the host team and matches where the home team is not the host team.

For each team, we will calculate the win percentage as the non-host team ad away team win percentage.

In [31]:
# Calculate the percentage of the host team's wins as the non-host team
non_host_team = fifa_wc[~host_team_mask]
non_host_team_win_percentage = non_host_team.groupby('home_team').agg({'home_win': 'sum', 'date': 'count'}).reset_index()
non_host_team_win_percentage.columns = ['team', 'wins', 'num_games']

# Calculate the away team win percentage
non_host_team_away_win_percentage = non_host_team.groupby('away_team').agg({'away_win': 'sum', 'date': 'count'}).reset_index()
non_host_team_away_win_percentage.columns = ['team', 'wins', 'num_games']

# Merge non-host team and away team win percentage
non_host_record = pd.merge(left=non_host_team_win_percentage, right=non_host_team_away_win_percentage, on=['team'], how='outer')

# Show the first five (5) rows of the non-host team win percentage
non_host_record.head()

team  wins_x  num_games_x  wins_y  num_games_y
0    algeria     1.0          6.0     2.0          7.0
1     angola     0.0          1.0     0.0          2.0
2  argentina    39.0         58.0     8.0         19.0
3  australia     3.0          7.0     1.0         12.0
4    austria     9.0         20.0     2.0          6.0

Now let's calculate the win percentage of non-host, we need to add up the total wins and divide them by the total number of games played.

In [32]:
# Calculate the win percentage of non-host team
non_host_record['non_host_win_percentage'] = (non_host_record['wins_x'] + non_host_record['wins_y']) / (non_host_record['num_games_x'] + non_host_record['num_games_y'])
non_host_record = non_host_record[['team', 'non_host_win_percentage']]

# Show the first five (5) rows of the non-host team win percentage
non_host_record.head()

team  non_host_win_percentage
0    algeria                 0.230769
1     angola                 0.000000
2  argentina                 0.610390
3  australia                 0.210526
4    austria                 0.423077

Now, we need to merge the non-host record to the host record to perform comparisons. We'll use a left merge to keep any teams that have hosted the world cup but never appeared as non-hosts.

In [33]:
# Merge the host team and non-host team win percentage
host_vs_non_host = pd.merge(left=host_team_win_percentage, right=non_host_record, left_on='team', right_on='team', how='left')

# Show the first ten (10) rows of the host vs non-host team win percentage
host_vs_non_host.head(10)

team  win_percentage  non_host_win_percentage
0  argentina        0.714286                 0.610390
1     brazil        0.615385                 0.708333
2      chile        0.666667                 0.280000
3    england        0.833333                 0.415385
4     france        0.888889                 0.540984
5    germany        0.857143                 0.604396
6      italy        0.833333                 0.516129
7      japan        0.500000                 0.238095
8     mexico        0.555556                 0.260870
9      qatar        0.000000                      NaN

Let's create a new column named `win_percentage_diff` that quantifies the difference in win percentage between the host and non-host teams.

In [34]:
# Create a new column win_percentage_diff
host_vs_non_host['win_percentage_diff'] = host_vs_non_host['win_percentage'] - host_vs_non_host['non_host_win_percentage']

# Sort the values by win_percentage_diff in ascending order
host_vs_non_host.sort_values(by='win_percentage_diff', ascending=False, inplace=True)

# Show the rows of the host vs non-host team win percentage
host_vs_non_host

team  win_percentage  non_host_win_percentage  \
17        uruguay        1.000000                 0.380000   
12    south korea        0.571429                 0.129032   
3         england        0.833333                 0.415385   
2           chile        0.666667                 0.280000   
4          france        0.888889                 0.540984   
6           italy        0.833333                 0.516129   
8          mexico        0.555556                 0.260870   
14         sweden        0.666667                 0.380952   
7           japan        0.500000                 0.238095   
5         germany        0.857143                 0.604396   
15    switzerland        0.500000                 0.352941   
11   south africa        0.333333                 0.200000   
10         russia        0.600000                 0.472222   
0       argentina        0.714286                 0.610390   
16  united states        0.250000                 0.266667   
1          brazil        0.615385                 0.708333   
13          spain        0.200000                 0.543860   
9           qatar        0.000000                      NaN   

    win_percentage_diff  
17             0.620000  
12             0.442396  
3              0.417949  
2              0.386667  
4              0.347905  
6              0.317204  
8              0.294686  
14             0.285714  
7              0.261905  
5              0.252747  
15             0.147059  
11             0.133333  
10             0.127778  
0              0.103896  
16            -0.016667  
1             -0.092949  
13            -0.343860  
9                   NaN

Based on our analysis, we only see two countries that have better record in the world cup as non-host than as host. These countries are **Brazil** abd **Qatar**, interestingly **Qatar** is te only country that has no non-host data to report.  

### Question 2. Who were the top goal scorers, individually and by team?

Next, we explore the `goalscorers.csv` dataset to determine the top goal scorers in the FIFA World Cup tournament, we will use the information to gain insights about:

- Top individual goal scorers
- Top scoring teams
- 1st half vs 2nd half team performances

We already loaded the `goalscorers.csv` dataset earlier in the notebook. Now let's clean and inspect the dataset.

In [35]:
# Clean the goalscorers.csv dataset
goalscorers = clean_data(df2)

# Show the first five (5) rows of the goalscorers dataset
goalscorers.head()

date  home_team away_team       team            scorer  minute  \
0  1916-07-02      chile   uruguay    uruguay   josé piendibene    44.0   
1  1916-07-02      chile   uruguay    uruguay  isabelino gradín    55.0   
2  1916-07-02      chile   uruguay    uruguay  isabelino gradín    70.0   
3  1916-07-02      chile   uruguay    uruguay   josé piendibene    75.0   
4  1916-07-06  argentina     chile  argentina     alberto ohaco     2.0   

   own_goal  penalty  
0     False    False  
1     False    False  
2     False    False  
3     False    False  
4     False    False

Initial observations: 

- Individual matches can be classified using the `date`, `home_team`, and `away_team` columns
- Each minute of goal was recorded in the `minute` column
- Goal and penalty information was recorded in the `scorer` and `penalty` columns, respectively.

Now, let's perform a left merge to create a dataset containing the goal scorers and the FIFA World Cup matches.

In [36]:
# Merge the goalscorers and FIFA World Cup datasets
fifa_wc_goalscorers = pd.merge(fifa_wc, goalscorers, on=['date', 'home_team', 'away_team'], how='left')

# Show the first five (5) rows of the FIFA World Cup goalscorers dataset
fifa_wc_goalscorers.head()

date home_team      away_team  home_score  away_score  country  \
0  1930-07-13   belgium  united states           0           3  uruguay   
1  1930-07-13   belgium  united states           0           3  uruguay   
2  1930-07-13   belgium  united states           0           3  uruguay   
3  1930-07-13    france         mexico           4           1  uruguay   
4  1930-07-13    france         mexico           4           1  uruguay   

   neutral  total_goals  win_margin         winner  home_win  away_win  \
0     True            3          -3  united states     False      True   
1     True            3          -3  united states     False      True   
2     True            3          -3  united states     False      True   
3     True            5           3         france      True     False   
4     True            5           3         france      True     False   

            team            scorer  minute own_goal penalty  
0  united states       bart mcghee    23.0    False   False  
1  united states        tom florie    45.0    False   False  
2  united states    bert patenaude    69.0    False   False  
3         france    lucien laurent    19.0    False   False  
4         france  marcel langiller    40.0    False   False

Let's explore the data by calculating which teams scored the most goals off penalties.

In [37]:
# Calculate the number of goals scored off penalties by team
penalty_goals = fifa_wc_goalscorers.groupby('team')['penalty'].sum()
penalty_goals = penalty_goals.sort_values(ascending=False).reset_index()
penalty_goals.columns = ['team', 'total_penalty_goals']

# Show the top 10 teams that scored the most goals off penalties
penalty_goals.head(10)

team total_penalty_goals
0        spain                  16
1       france                  15
2    argentina                  12
3      germany                  12
4      england                  12
5  netherlands                  10
6       brazil                  10
7       mexico                  10
8     portugal                   9
9        italy                   8

Now, let's the top 10 teams with the most goals scored into their own goal in the FIFA World Cup tournament.

In [38]:
# Calculate the number of goals scored by team
total_goals = fifa_wc_goalscorers.groupby('team')['own_goal'].sum()
total_goals = total_goals.sort_values(ascending=False).reset_index()
total_goals.columns = ['team', 'total_total_goals']

# Show the top 10 teams that scored the most goals
total_goals.head(10)

team total_total_goals
0         france                 6
1        germany                 4
2          italy                 4
3       portugal                 3
4  united states                 3
5        belgium                 2
6        austria                 2
7         russia                 2
8          spain                 2
9       paraguay                 2

Let's calculate the all-time leading goal scorers in the FIFA World Cup tournament.

In [39]:
# Calculate the all-time leading goal scorers
leading_goal_scorers = fifa_wc_goalscorers.groupby('scorer')['date'].count()
leading_goal_scorers = leading_goal_scorers.sort_values(ascending=False).reset_index()
leading_goal_scorers.columns = ['scorer', 'total_goals']

# Show the top 10 all-time leading goal scorers
leading_goal_scorers.head(10)

scorer  total_goals
0    miroslav klose           16
1           ronaldo           15
2       gerd müller           14
3      lionel messi           13
4     just fontaine           13
5     kylian mbappé           12
6              pelé           12
7  jürgen klinsmann           11
8     sándor kocsis           11
9  teófilo cubillas           10

Based on the analysis, **Miroslav Klose** is the all-time leading goal scorer in the FIFA World Cup tournament with **16** goals. We also saw some very famous retired players such as **Klose** and **Ronaldo** and current active players such as **Mbappe** and **Messi**.

Then now, let's calculate the top individual goal scorers for each team.

We need to create a function that counts the number of goals scored by individual players for each team and extract the top scorer. 

In [40]:
# Calculate the top individual goal scorers for each team
def get_top_scorer(column):
    value_counts = column.value_counts()
    top_scorer = value_counts.idxmax()
    goals_scored = value_counts.max()
    return str(top_scorer) + ',' + str(goals_scored)

top_scorers = fifa_wc_goalscorers.groupby('team')['scorer'].agg(get_top_scorer).reset_index()

# Show the top individual goal scorers for each team
top_scorers

team             scorer
0                algeria      salah assad,2
1                 angola     flávio amado,1
2              argentina    lionel messi,13
3              australia       tim cahill,5
4                austria     erich probst,6
..                   ...                ...
73  united arab emirates    khalid ismaïl,1
74         united states   landon donovan,5
75               uruguay     óscar míguez,8
76                 wales   ivor allchurch,2
77            yugoslavia  dražan jerković,4

[78 rows x 2 columns]

Let's split the scorer column, `top_scorer` and `goals` into two separate columns. 

In [41]:
# Perform the split operation once and assign the results to two new columns
top_scorers[['top_scorer', 'goals']] = top_scorers['scorer'].str.split(',', expand=True)

# Drop the 'scorer' column
top_scorers.drop(columns=['scorer'], inplace=True)

# Show the top individual goal scorers for each team
top_scorers

team       top_scorer goals
0                algeria      salah assad     2
1                 angola     flávio amado     1
2              argentina     lionel messi    13
3              australia       tim cahill     5
4                austria     erich probst     6
..                   ...              ...   ...
73  united arab emirates    khalid ismaïl     1
74         united states   landon donovan     5
75               uruguay     óscar míguez     8
76                 wales   ivor allchurch     2
77            yugoslavia  dražan jerković     4

[78 rows x 3 columns]

Based on the analysis, **Ronaldo** from Brazil team is the all-time leading scorer in FIFA World Cup.

### Question 3. Are the teams who perform better in the first (1st) or second (2nd) half of the match more likely to win the match?

We might be interested in comparing each team's performance in the 1st and 2nd half of their matches. This analysis can help us determine if teams who perform better in the 1st or 2nd half of the match are more likely to win the match.

First let's see the summary statistics of the `minute` column.

In [42]:
# Display the summary statistics of the 'minute' column
fifa_wc_goalscorers['minute'].describe()

count    2720.000000
mean       51.581250
std        27.535554
min         1.000000
25%        28.000000
50%        53.000000
75%        75.000000
max       120.000000
Name: minute, dtype: float64

Now, let's classify the goals scored in the 1st and 2nd half of the match. Remember that the 1st half (0-45 minutes) and the 2nd half (45-90 minutes) of the match.

NOTE: It is important to consider that it is not a perfect measure since extra-time maybe added at the end of each half due to breaks, player injuries, or other reasons.

In [43]:
# Classify the goals scored in the 1st and 2nd half of the match
def classify_goal_time(row):
    if row['minute'] <= 45.0:
        output = '1st half'
    elif row['minute'] <= 90.0:
        output = '2nd half'
    else:
        output = 'Overtime'
        
    return output

In [44]:
# Apply the classify_goal_time function to the FIFA World Cup goalscorers dataset
fifa_wc_goalscorers['goal_time'] = fifa_wc_goalscorers.apply(classify_goal_time, axis=1)

# Show the first five (5) rows of the FIFA World Cup goalscorers dataset
fifa_wc_goalscorers.head()

date home_team      away_team  home_score  away_score  country  \
0  1930-07-13   belgium  united states           0           3  uruguay   
1  1930-07-13   belgium  united states           0           3  uruguay   
2  1930-07-13   belgium  united states           0           3  uruguay   
3  1930-07-13    france         mexico           4           1  uruguay   
4  1930-07-13    france         mexico           4           1  uruguay   

   neutral  total_goals  win_margin         winner  home_win  away_win  \
0     True            3          -3  united states     False      True   
1     True            3          -3  united states     False      True   
2     True            3          -3  united states     False      True   
3     True            5           3         france      True     False   
4     True            5           3         france      True     False   

            team            scorer  minute own_goal penalty goal_time  
0  united states       bart mcghee    23.0    False   False  1st half  
1  united states        tom florie    45.0    False   False  1st half  
2  united states    bert patenaude    69.0    False   False  2nd half  
3         france    lucien laurent    19.0    False   False  1st half  
4         france  marcel langiller    40.0    False   False  1st half

Now, that we already classified each goal into `1st half` and `2nd half`, let's calculate the number of goals scored in the 1st and 2nd half of the match.

In [45]:
# Calculate the number of goals scored in the 1st and 2nd half of the match by team
team_goal_time = pd.pivot_table(fifa_wc_goalscorers, values='scorer', index='team', columns='goal_time', aggfunc='count')

# Show the first five (5) rows of the team goal time
team_goal_time.head()

goal_time  1st half  2nd half  Overtime
team                                   
algeria         7.0       5.0       1.0
angola          NaN       1.0       NaN
argentina      69.0      78.0       5.0
australia       7.0      10.0       NaN
austria        18.0      23.0       2.0

Notice that some teams may not have scored any goals in the 1st or 2nd half of the match. Let's fill the missing values with `NaN`.

Now, let's convert the counts to percentages to compare the 1st, 2nd half, and overtime performances of each team. We need to divide the number of goals in each half, respectively, by the total number of goals scored by each team.

In [46]:
# Convert the counts to percentages
team_goal_time['total_goals'] = team_goal_time[['1st half', '2nd half', 'Overtime']].sum(axis=1)

team_goal_time['1st half %'] = team_goal_time['1st half'] / team_goal_time['total_goals']
team_goal_time['2nd half %'] = team_goal_time['2nd half'] / team_goal_time['total_goals']
team_goal_time['Overtime %'] = team_goal_time['Overtime'] / team_goal_time['total_goals']

# Show the first five (5) rows of the team goal time
team_goal_time.head()

goal_time  1st half  2nd half  Overtime  total_goals  1st half %  2nd half %  \
team                                                                           
algeria         7.0       5.0       1.0         13.0    0.538462    0.384615   
angola          NaN       1.0       NaN          1.0         NaN    1.000000   
argentina      69.0      78.0       5.0        152.0    0.453947    0.513158   
australia       7.0      10.0       NaN         17.0    0.411765    0.588235   
austria        18.0      23.0       2.0         43.0    0.418605    0.534884   

goal_time  Overtime %  
team                   
algeria      0.076923  
angola            NaN  
argentina    0.032895  
australia         NaN  
austria      0.046512

Let's see which teams scored the most goals in the 2nd half of the match.

In [47]:
# Calculate the number of goals scored in the 2nd half of the match by team
second_half_goals = team_goal_time.sort_values('2nd half %', ascending=False)

# Show the top 10 teams that scored the most goals in the 2nd half of the match
second_half_goals.head(10)

goal_time    1st half  2nd half  Overtime  total_goals  1st half %  \
team                                                                 
qatar             NaN       1.0       NaN          1.0         NaN   
el salvador       NaN       1.0       NaN          1.0         NaN   
haiti             NaN       2.0       NaN          2.0         NaN   
kuwait            NaN       2.0       NaN          2.0         NaN   
bolivia           NaN       1.0       NaN          1.0         NaN   
angola            NaN       1.0       NaN          1.0         NaN   
iraq              NaN       1.0       NaN          1.0         NaN   
israel            NaN       1.0       NaN          1.0         NaN   
norway            1.0       6.0       NaN          7.0    0.142857   
costa rica        4.0      18.0       NaN         22.0    0.181818   

goal_time    2nd half %  Overtime %  
team                                 
qatar          1.000000         NaN  
el salvador    1.000000         NaN  
haiti          1.000000         NaN  
kuwait         1.000000         NaN  
bolivia        1.000000         NaN  
angola         1.000000         NaN  
iraq           1.000000         NaN  
israel         1.000000         NaN  
norway         0.857143         NaN  
costa rica     0.818182         NaN

Filter the teams who have scored total goals greater that the 25th percentile.

In [48]:
# Filter the teams who have scored total goals greater that the 25th percentile
team_goal_time = team_goal_time[team_goal_time['total_goals'] > team_goal_time['total_goals'].quantile(0.25)]

# Show the first five (5) rows of the team goal time
team_goal_time.head()

goal_time  1st half  2nd half  Overtime  total_goals  1st half %  2nd half %  \
team                                                                           
algeria         7.0       5.0       1.0         13.0    0.538462    0.384615   
argentina      69.0      78.0       5.0        152.0    0.453947    0.513158   
australia       7.0      10.0       NaN         17.0    0.411765    0.588235   
austria        18.0      23.0       2.0         43.0    0.418605    0.534884   
belgium        28.0      36.0       5.0         69.0    0.405797    0.521739   

goal_time  Overtime %  
team                   
algeria      0.076923  
argentina    0.032895  
australia         NaN  
austria      0.046512  
belgium      0.072464

It looks like, we have more data to better evaluate each team's performance in the 1st and 2nd half of the match. 

We now compare their performance. Calculate each team's overall win percentage so that we can compare it to their 1st and 2nd half performance.

We need to combine each team's record as both the home and away team to get their overall win percentage.

In [49]:
# Calculate home team wins and losses
home_record = pd.pivot_table(fifa_wc_goalscorers,
                             index='home_team',
                             columns='home_win',
                             values='date',
                             aggfunc='count').reset_index()

home_record.columns = ['team', 'home_losses', 'home_wins']

# Show the first five (5) rows of the home record
home_record.head()

team  home_losses  home_wins
0    algeria         10.0        5.0
1     angola          1.0        NaN
2  argentina         54.0      141.0
3  australia         11.0        8.0
4    austria         35.0       30.0

In [50]:
# Calculate away team wins and losses
away_record = pd.pivot_table(fifa_wc_goalscorers,
                             index='away_team',
                             columns='away_win',
                             values='date',
                             aggfunc='count').reset_index()

away_record.columns = ['team', 'away_losses', 'away_wins']

# Show the first five (5) rows of the away record
away_record.head()

team  away_losses  away_wins
0    algeria          9.0        9.0
1     angola          3.0        NaN
2  argentina         43.0       21.0
3  australia         35.0        1.0
4    austria         10.0       15.0

We have missing, `NaN`, values corresponding to 0 wins.

Now, let's merge the home and away records to calculate the overall win percentage for each team.

In [51]:
# Merge the home and away records
overall_record = pd.merge(home_record, away_record, left_on='team', right_on='team', how='inner')
overall_record['total_wins'] = overall_record[['home_wins', 'away_wins']].sum(axis=1)
overall_record['total_losses'] = overall_record[['home_losses', 'away_losses']].sum(axis=1)
overall_record['win %'] = overall_record['total_wins'] / (overall_record['total_wins'] + overall_record['total_losses'])

# Show the first ten (10) rows of the overall record
overall_record[['team', 'total_wins', 'total_losses', 'win %']].head(10)

team  total_wins  total_losses     win %
0                 algeria        14.0          19.0  0.424242
1                  angola         0.0           4.0  0.000000
2               argentina       162.0          97.0  0.625483
3               australia         9.0          46.0  0.163636
4                 austria        45.0          45.0  0.500000
5                 belgium        60.0          85.0  0.413793
6                 bolivia         0.0          22.0  0.000000
7  bosnia and herzegovina         4.0           4.0  0.500000
8                  brazil       267.0          87.0  0.754237
9                bulgaria        11.0          67.0  0.141026

Now, let's merge the overall record with the team goal time dataset to compare each team's performance in the 1st and 2nd half of the match.

In [52]:
# Merge the overall record with the team goal time dataset
goals_record = pd.merge(overall_record, team_goal_time, left_on='team', right_on='team', how='inner')
goals_record = goals_record[['team', 'total_wins', 'total_losses', 'win %', '1st half %', '2nd half %', 'Overtime %']]

# Show the first five (5) rows of the goals record
goals_record.head()

team  total_wins  total_losses     win %  1st half %  2nd half %  \
0    algeria        14.0          19.0  0.424242    0.538462    0.384615   
1  argentina       162.0          97.0  0.625483    0.453947    0.513158   
2  australia         9.0          46.0  0.163636    0.411765    0.588235   
3    austria        45.0          45.0  0.500000    0.418605    0.534884   
4    belgium        60.0          85.0  0.413793    0.405797    0.521739   

   Overtime %  
0    0.076923  
1    0.032895  
2         NaN  
3    0.046512  
4    0.072464

Let's see which teams scored most of their goals in the 1st half of the match.

In [53]:
# Calculate the number of goals scored in the 1st half of the match by team
first_half_goals = goals_record.sort_values('1st half %', ascending=False)

# Show the top 10 teams that scored the most goals in the 1st half of the match
first_half_goals.head(10)

team  total_wins  total_losses     win %  1st half %  \
28       north korea         1.0          26.0  0.037037    0.666667   
48            turkey        17.0          20.0  0.459459    0.650000   
41            serbia         4.0          41.0  0.088889    0.625000   
27           nigeria        14.0          41.0  0.254545    0.565217   
40           senegal        14.0          19.0  0.424242    0.562500   
39          scotland        17.0          51.0  0.250000    0.560000   
25           morocco        14.0          37.0  0.274510    0.550000   
29  northern ireland         5.0          32.0  0.135135    0.538462   
0            algeria        14.0          19.0  0.424242    0.538462   
14           ecuador        11.0          18.0  0.379310    0.500000   

    2nd half %  Overtime %  
28    0.333333         NaN  
48    0.300000    0.050000  
41    0.375000         NaN  
27    0.434783         NaN  
40    0.375000    0.062500  
39    0.440000         NaN  
25    0.450000         NaN  
29    0.384615    0.076923  
0     0.384615    0.076923  
14    0.500000         NaN

Get the average or mean value of goals at each half of the match.

In [54]:
print('Average goals scored in the 1st half:', first_half_goals['1st half %'].head(10).mean())
print('Average goals scored in the 2nd half:', first_half_goals['2nd half %'].head(10).mean())

Average goals scored in the 1st half: 0.5756307134894091
Average goals scored in the 2nd half: 0.39773467112597544


Get the average win percentage of the teams.

In [55]:
print('Average win percentage of the teams:', first_half_goals['win %'].head().mean())

Average win percentage of the teams: 0.25283465283465284
